compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt

```diff
@@ -818,23 +836,17 @@ class FirDataFlowAnalyzer(private val components: FirAbstractBodyResolveTransfor
     private fun getOrCreateSyntheticVariable(fir: FirElement): SyntheticDataFlowVariable =
         variableStorage.getOrCreateNewSyntheticVariable(fir)
 
-    private fun FirElement.unwrapWhenSubjectExpression(): FirElement = if (this is FirWhenSubjectExpression) {
-        val whenExpression = whenSubject.whenExpression
-        whenExpression.subjectVariable
-            ?: whenExpression.subject
-            ?: throw IllegalStateException("Subject or subject variable must be not null")
-    } else {
-        this
-    }
-
     private fun getOrCreateRealVariable(fir: FirElement): RealDataFlowVariable? {
         @Suppress("NAME_SHADOWING")
-        val fir = fir.unwrapWhenSubjectExpression()
-        if (fir is FirThisReceiverExpressionImpl) {
-            return variableStorage.getOrCreateNewThisRealVariable(fir.calleeReference.boundSymbol ?: return null)
+        val fir = fir.unwrapWhenSubjectExpression().unwrapSmartcast()
+        val symbol = fir.resolvedSymbol ?: return null
+        return when {
+            fir is FirThisReceiverExpressionImpl -> variableStorage.getOrCreateNewThisRealVariable(symbol)
+            symbol is FirVariableSymbol<*> && symbol.fir.run {
+                this is FirProperty && isLocal || this is FirValueParameter
+            } -> variableStorage.getOrCreateNewRealVariable(symbol).variableUnderAlias
```

Comment:
@demiurg906 - I made this change after I noticed an issue with the following code (this is also in `testData/resolve/smartcasts/notBoundSmartcasts.kt`):
```kt
fun test_3(d1: D, d2: D) {
    val a = d1?.any
    val b = d2?.any
    a as A
    a.foo()
    b as B
    b.foo()
}
```

The issue was we would be storing a `RealDataFlowVariable` for the non-local property symbol, and therefore erroneously smartcasting local variables that access the same property, even for different objects.

This change then broke some tests in `testData/resolve/smartcasts/nullability.kt`. After analyzing those failures, I noticed some incorrectness in `getRealVariablesForSafeCallChain()` -- see the comment there. It doesn't seem like a simple fix. I verified against the current frontend as well. The current frontend would reject certain smartcasts that FIR frontend would accept -- see the changes in the `nullability.kt`.

Let me know if my understanding of this is correct!

Reply:
It's not correct change, because we can smartcast final member properties:
```
class A {
    val x: Any = ""
}

fun test(a: A) {
    if (a.x is String) {
        a.x.length // OK in old frontend
    }
}
```

Reply:
Please, remove this change (but keep tests, they are good)

Reply:
I planned to start rewrite `DataFlowAnalyzer` tomorrow, so you tests and exploration will be helpful, thanks

Reply:
I understand that it's possible to smartcast final member properties, but the current implementation is broken for the case I described above (type info is associated with the symbol). This change _does_ break it even further though, which would explain the increase in errors you saw.

Okay, I will revert this part of the changes for now.